interp/src/interpreter/interpret_group.rs

```diff
@@ -1,357 +1,479 @@
+        }
+    }
+
+    fn do_tick(&mut self) -> Vec<*const ir::Port> {
+        self.backing_env.clk += 1;
+
+        let mut new_vals = vec![];
+        let mut w_env = std::mem::take(&mut self.working_env);
+
+        self.working_env = w_env
+            .drain()
+            .filter_map(|(port, val)| match val {
+                OutputValue::ImmediateValue(iv) => {
+                    self.backing_env.insert(port, iv);
+                    None
+                }
+                out @ OutputValue::PulseValue(_)
+                | out @ OutputValue::LockedValue(_) => {
+                    let old_val = out.get_val().clone();
+
+                    match out.do_tick() {
+                        OutputValue::ImmediateValue(iv) => {
+                            if iv != old_val {
+                                new_vals.push(port)
+                            }
+                            self.backing_env.insert(port, iv);
+                            None
+                        }
+                        v @ OutputValue::LockedValue(_) => Some((port, v)),
+                        OutputValue::PulseValue(pv) => {
+                            if *pv.get_val() != old_val {
+                                new_vals.push(port);
+                            }
+                            Some((port, pv.into()))
+                        }
+                    }
+                }
+            })
+            .collect();
+        new_vals
+    }
+
+    fn collapse_env(mut self) -> Environment {
+        let working_env = self.working_env;
+
+        for (port, v) in working_env {
+            match v {
+                OutputValue::ImmediateValue(iv) => {
+                    self.backing_env.insert(port, iv)
+                }
+                OutputValue::LockedValue(tlv) => {
+                    if tlv.unlockable() {
+                        let iv = tlv.unlock();
+                        self.backing_env.insert(port, iv);
+                    } else {
+                        panic!("Group is done with an invalid value?")
+                    }
+                }
+                OutputValue::PulseValue(v) => {
+                    self.backing_env.insert(port, v.take_val())
                 }
-                map.insert(cb.clone_name(), ports);
             }
-            _ => panic!("component"),
         }
+        self.backing_env
+    }
+}
+
+// possibly #[inline] here later? Compiler probably knows to do that already
+fn get_done_port(group: &ir::Group) -> RRC<ir::Port> {
+    group.get(&"done")
+}
+
+fn grp_is_done(done: OutputValueRef) -> bool {
+    match done {
+        OutputValueRef::ImmediateValue(v) => v.as_u64() == 1,
+        OutputValueRef::LockedValue(_) => false,
+        OutputValueRef::PulseValue(v) => v.get_val().as_u64() == 1,
     }
-    map
 }
 
 /// Evaluates a group, given an environment.
 pub fn interpret_group(
     group: &ir::Group,
+    _continuous_assignments: &[ir::Assignment],
```

Comment:
Is this meant to be used in the future? If so, `TODO`.

Reply:
yes, I'll make a note